### PR TITLE
[qt] feat: Implement correct touch handling

### PR DESCRIPTION
**Pull request by Ferenc-**
* This adds the features for https://github.com/organicmaps/organicmaps/issues/5306
  - two finger pinch rotate gesture
  - pan
  - swipe

* Makes the previous two finger pinch zoom gesture feautre smoother and the zooming is now proportional to the gesture motion.

### CODE REVIEW

File: qt/draw_widget.cpp [hunk 1]
```diff
@@ -195,6 +214,35 @@ void DrawWidget::initializeGL()
     m_screenshoter->Start();
 }
 
+bool DrawWidget::event(QEvent * event)
+{
+  if (!isQtTouchEvent(event))
+    return QOpenGLWidget::event(event);
```
Comment (vng) [inline qt/draw_widget.cpp:220]: Simpler is better here:
- remove isQtTouchEvent
- add df::TouchEvent::NONE and return it in qtTouchEventTypeToDfTouchEventType
- call qtTouchEventTypeToDfTouchEventType here once and check for df::TouchEvent::NONE

Comment (Ferenc-) [reply qt/draw_widget.cpp:215]: Done

File: drape_frontend/user_event_stream.hpp [hunk 2]
```diff
@@ -72,7 +72,8 @@ class TouchEvent : public UserEvent
     TOUCH_DOWN,
     TOUCH_MOVE,
     TOUCH_UP,
-    TOUCH_CANCEL
+    TOUCH_CANCEL,
+    TOUCH_NONE
```
Comment (vng) [inline drape_frontend/user_event_stream.hpp:76]: nit: I'd prefer to put NONE on the first place.

File: qt/draw_widget.cpp [hunk 3]
```diff
@@ -78,6 +78,19 @@ void DrawMwmBorder(df::DrapeApi & drapeApi, std::string const & mwmName,
...
+{
+  switch (qEventType)
+  {
+    case QEvent::TouchBegin: return df::TouchEvent::TOUCH_DOWN;
+    case QEvent::TouchEnd: return df::TouchEvent::TOUCH_UP;
+    case QEvent::TouchUpdate: return df::TouchEvent::TOUCH_MOVE;
+    case QEvent::TouchCancel: return df::TouchEvent::TOUCH_CANCEL;
+    default: return df::TouchEvent::TOUCH_NONE;
+  }
+  UNREACHABLE();
```
Comment (vng) [inline qt/draw_widget.cpp:92]: nit: No need here.

**Review by vng — APPROVED** [review]

File: drape_frontend/user_event_stream.hpp (hunk 2 above)
Comment (Ferenc-) [reply drape_frontend/user_event_stream.hpp]: Done

File: qt/draw_widget.cpp (hunk 3 above)
Comment (Ferenc-) [reply qt/draw_widget.cpp]: Done

**Comment by vng** [comment]
@Ferenc- 

I constantly get this after merging on my Mac:
```
4) ASSERT FAILED
drape_frontend/user_event_stream.cpp:897
CHECK(GetValidTouchesCount(m_touches) == 2) 1 2 
```

**Comment by Ferenc-** [comment]
> LMB down and drag with another finger on a touchpad.

On Linux this doesn''t seem reproducible. Which version of MacOS is this?

**Comment by vng** [comment]
The latest one, 15.1
More important here is that we still use Qt 6.6.3m which one is yours?

**Comment by RedAuburn** [comment]
Having tested it, it's no different at all to before on GNOME 🤔
I'm using Silverblue and running OM from a container, but it shouldn't be an issue. Does it definitely work on KDE?

**Comment by Ferenc-** [comment]
> Does it definitely work on KDE?

@RedAuburn It does definitely work on Gnome as well, and also on Fedora 41 specifically. What kind of touchscreen do you have?

**Comment by RedAuburn** [comment]
ah, i was under the impression that this would add support for trackpad gestures as well, unfortunately i don't have a touchscreen to test with

**Pull request merged**